projects/timeline_semaphore/main.cpp

```diff
@@ -0,0 +1,328 @@
+
+        PPX_CHECKED_CALL(GetDevice()->CreateTextureFont(&createInfo, &mRoboto));
+    }
+
+    // Text draw
+    {
+        grfx::ShaderModulePtr VS;
+        grfx::ShaderModulePtr PS;
+
+        std::vector<char> bytecode = LoadShader("basic/shaders", "TextDraw.vs");
+        PPX_ASSERT_MSG(!bytecode.empty(), "VS shader bytecode load failed");
+        grfx::ShaderModuleCreateInfo shaderCreateInfo = {static_cast<uint32_t>(bytecode.size()), bytecode.data()};
+        PPX_CHECKED_CALL(GetDevice()->CreateShaderModule(&shaderCreateInfo, &VS));
+
+        bytecode = LoadShader("basic/shaders", "TextDraw.ps");
+        PPX_ASSERT_MSG(!bytecode.empty(), "PS shader bytecode load failed");
+        shaderCreateInfo = {static_cast<uint32_t>(bytecode.size()), bytecode.data()};
+        PPX_CHECKED_CALL(GetDevice()->CreateShaderModule(&shaderCreateInfo, &PS));
+
+        grfx::TextDrawCreateInfo createInfo = {};
+        createInfo.pFont                    = mRoboto;
+        createInfo.maxTextLength            = 4096;
+        createInfo.VS                       = {VS.Get(), "vsmain"};
+        createInfo.PS                       = {PS.Get(), "psmain"};
+        createInfo.renderTargetFormat       = GetSwapchain()->GetColorFormat();
+
+        PPX_CHECKED_CALL(GetDevice()->CreateTextDraw(&createInfo, &mDynamicText));
+
+        GetDevice()->DestroyShaderModule(VS);
+        GetDevice()->DestroyShaderModule(PS);
+    }
+}
+
+void ProjApp::Render()
+{
+    PerFrame& frame = mPerFrame[0];
+
+    grfx::SwapchainPtr swapchain = GetSwapchain();
+
+    uint32_t imageIndex = UINT32_MAX;
+    PPX_CHECKED_CALL(swapchain->AcquireNextImage(UINT64_MAX, frame.imageAcquiredSemaphore, frame.imageAcquiredFence, &imageIndex));
+
+    // Wait for and reset image acquired fence
+    PPX_CHECKED_CALL(frame.imageAcquiredFence->WaitAndReset());
+
+    // Wait for previous frame's render to complete (GPU signal to CPU wait)
+    PPX_CHECKED_CALL(frame.timelineSemaphore->Wait(frame.timelineValue));
+
+    // Spawn a thread that will spawn other threads to signal values on the CPU
+    std::unique_ptr<std::thread> spawnerThread;
+    {
+        const uint32_t   kNumThreads = 4;
+        grfx::Semaphore* pSemaphore  = frame.timelineSemaphore;
+
+        // Normally, we increment after a wait and before the next signal so we need to add 1.
+        const uint64_t startSignalValue = frame.timelineValue + 1;
+
+        spawnerThread = std::unique_ptr<std::thread>(
+            new std::thread([kNumThreads, pSemaphore, startSignalValue]() {
+                std::vector<std::unique_ptr<std::thread>> signalThreads;
+
+                // Create signaling threads
+                for (uint32_t i = 0; i < kNumThreads; ++i) {
+                    const uint64_t signalValue = startSignalValue + i;
+
+                    auto signalThread = std::unique_ptr<std::thread>(
+                        new std::thread([pSemaphore, signalValue] {
```

Comment:
are those guaranteed to happen in order (so that the signal value is increasing)? If not, what happens if they are nto in the increasing ordeR?

Reply:
They're not in order on purpose. The `forceMonotonicValue=true` guards against signaling a value that's under what's presently in the semaphore.